drivers/mtd/mtd.c

```diff
@@ -42,12 +43,57 @@ int mtd_read(mtd_dev_t *mtd, void *dest, uint32_t addr, uint32_t count)
         return -ENODEV;
     }
 
-    if (mtd->driver->read) {
-        return mtd->driver->read(mtd, dest, addr, count);
+    /* page size is always a power of two */
+    const uint32_t page_shift = 32 - __builtin_clz(mtd->page_size) - 1;
+    const uint32_t page_mask = mtd->page_size - 1;
+
+    return mtd_read_page(mtd, dest, addr >> page_shift, addr & page_mask, count);
+}
+
+int mtd_read_page(mtd_dev_t *mtd, void *dest, uint32_t page, uint32_t offset,
+                  uint32_t count)
+{
+    if (!mtd || !mtd->driver) {
+        return -ENODEV;
     }
-    else {
-        return -ENOTSUP;
+
+    if (mtd->driver->read_page == NULL) {
+        /* TODO: remove when all backends implement read_page */
+        if (mtd->driver->read) {
+            return mtd->driver->read(mtd, dest, mtd->page_size * page + offset, count);
+        } else {
+            return -ENOTSUP;
+        }
     }
+
+    char *_dst = dest;
+
+    /* page size is always a power of two */
+    const uint32_t page_shift = 32 - __builtin_clz(mtd->page_size) - 1;
+    const uint32_t page_mask = mtd->page_size - 1;
+
+    page  += offset >> page_shift;
+    offset = offset & page_mask;
+
+    while (count) {
+        int read = mtd->driver->read_page(mtd, _dst, page, offset, count);
```

Comment:
Bit of a nitpick, but this clashes heavily in my mind with `read()`. Care to rename this variable to something slightly different?

Reply:
Thanks! :)